server/server.go

```diff
@@ -127,113 +120,91 @@ func (s *Server) HandleEvent(ctx context.Context, e lookout.Event) error {
-	ctx, logger := ctxlog.WithLogFields(ctx, log.Fields{
-		"provider": e.Provider,
-	})
-	logger.Infof("processing pull request")
+func (s *Server) AnalyzeAndComment(ctx context.Context, e lookout.Event, safePosting bool) error {
+	comments, err := s.analyze(ctx, e, safePosting)
+	if err == nil {
+		if err := s.post(ctx, e, comments, safePosting); err != nil {
+			s.status(ctx, e, lookout.ErrorAnalysisStatus)
+			err = fmt.Errorf("posting analysis failed: %s", err)
+		} else {
+			s.status(ctx, e, lookout.SuccessAnalysisStatus)
+		}
+	}
+	return err
+}
 
+func (s *Server) analyze(ctx context.Context, e lookout.Event, safePosting bool) ([]lookout.AnalyzerComments, error) {
+	ctxlog.Get(ctx).Infof("processing event type %d", e.Type())
+
+	var comments []lookout.AnalyzerComments
 	if err := e.Validate(); err != nil {
-		return err
+		return comments, err
 	}
 
 	conf, err := s.getConfig(ctx, e)
 	if err != nil {
-		return err
+		return comments, err
 	}
 
 	s.status(ctx, e, lookout.PendingAnalysisStatus)
 
-	send := func(
-		ctx context.Context,
-		a lookout.AnalyzerClient,
-		settings map[string]interface{},
-	) ([]*lookout.Comment, error) {
-		st := grpchelper.ToPBStruct(settings)
-		if st != nil {
-			e.Configuration = *st
-		}
-
-		if s.analyzerReviewTimeout > 0 {
-			var cancel context.CancelFunc
-			ctx, cancel = context.WithTimeout(ctx, s.analyzerReviewTimeout)
-			defer cancel()
-		}
-
-		resp, err := a.NotifyReviewEvent(ctx, e)
-		if err != nil {
-			return nil, err
-		}
-		return resp.Comments, nil
-	}
-	comments, err := s.concurrentRequest(ctx, conf, send, grpcErrorMessages[pb.ReviewEventType])
-	if err != nil {
-		return err
-	}
-
-	if err := s.post(ctx, e, comments, safePosting); err != nil {
-		s.status(ctx, e, lookout.ErrorAnalysisStatus)
-		return fmt.Errorf("posting analysis failed: %s", err)
-	}
-
-	s.status(ctx, e, lookout.SuccessAnalysisStatus)
-
-	return nil
+	send := s.getSender(e)
+	return s.concurrentRequest(ctx, conf, send, grpcErrorMessages[pb.ReviewEventType])
 }
 
-// HandlePush sends request to analyzers concurrently
-func (s *Server) HandlePush(ctx context.Context, e *lookout.PushEvent, safePosting bool) error {
-	ctx, logger := ctxlog.WithLogFields(ctx, log.Fields{
-		"provider": e.Provider,
-	})
-	logger.Infof("processing push")
-
-	if err := e.Validate(); err != nil {
-		return err
+func (s *Server) getSender(e lookout.Event) reqSent {
+	type analyzerNotifier struct {
+		notify  func(context.Context, lookout.AnalyzerClient) (*pb.EventResponse, error)
+		timeout time.Duration
 	}
 
-	conf, err := s.getConfig(ctx, e)
-	if err != nil {
-		return err
+	getAnalyzerNotifier := func(ctx context.Context, settings map[string]interface{}) (analyzerNotifier, error) {
+		st := grpchelper.ToPBStruct(settings)
+		switch ev := e.(type) {
+		case *lookout.ReviewEvent:
+			if st != nil {
+				ev.Configuration = *st
+			}
+			return analyzerNotifier{
+				func(ctx context.Context, a lookout.AnalyzerClient) (*pb.EventResponse, error) {
```

Comment:
We try to avoid structs initialized with positional arguments because they're harder to read and can cause problems in some cases (like changing the order of properties having the same type, or adding new properties).

imho I'd prefer:
```go
return analyzerNotifier{
  notify: func...
  timeot: s.analyzerReviewTimeout
}, nil
```

Reply:
Agree